praxis/contrib/gpu/scripts_gpu/te_helper.py

```diff
@@ -112,19 +115,85 @@ def _fprop(
 
     @staticmethod
     def set_layer_params_to_stack_transformer(stacked_transformer_obj, _, layer_id):
+        enable_sp = bool(int(os.environ.get('ENABLE_TE_SP', 0)))
         te_transformer_tpl = pax_fiddle.Config(te_praxis.TransformerLayer,
             name=f'layer_{layer_id}',
-            layernorm_type='layernorm',
-            zero_centered_gamma = True,
-            mlp_activations=('gelu',),
-            use_bias=True,
-            self_attn_mask_type='causal',
             enable_relative_embedding=False,
-            scaled_query_init=False,
-            scale_attn_logits=True,
+            enable_sequence_parallel=enable_sp,
             transpose_batch_sequence=False
         )
 
+        def update_ln_te_tpl(te_tpl, transformer_layer_tpl):
+            # TE requires all normalization are the same
+            assert transformer_layer_tpl.ln_tpl == transformer_layer_tpl.tr_fflayer_tpl.ln_tpl
+            ln_tpl = transformer_layer_tpl.ln_tpl
+            if issubclass(ln_tpl.cls, normalizations.LayerNorm):
+                te_tpl.layernorm_type = 'layernorm'
+                assert ln_tpl.use_scale
+                assert ln_tpl.use_bias
+            elif issubclass(ln_tpl.cls, normalizations.RmsNorm):
+                te_tpl.layernorm_type = 'rmsnorm'
+            else:
+                raise ValueError(f'Unsupported {ln_tpl.cls=}, LayerNorm, RmsNorm are supported.')
+            te_tpl.zero_centered_gamma = not ln_tpl.direct_scale
+            te_tpl.layernorm_epsilon = ln_tpl.epsilon
+            return te_tpl
+
+        def update_ff_te_tpl(te_tpl, ff_layer_tpl):
+            if issubclass(ff_layer_tpl.activation_tpl.cls, activations.Identity):
+                mlp_activations = ('linear',)
+            else:
+                mlp_activations = (ff_layer_tpl.activation_tpl.cls.__name__.lower(),)
+            if ff_layer_tpl.use_gated_activation:
+                mlp_activations += ('linear',)
+            te_tpl.mlp_activations = mlp_activations
+            return te_tpl
+
+        def update_attn_te_tpl(te_tpl, attn_tpl):
+            # TODO(rewang): rope
+            if issubclass(attn_tpl.cls, attentions.DotProductAttention):
+                # Check the DotProductAttention parameters are aligned to TE's attention
+                assert attn_tpl.internal_enable_query_scale or attn_tpl.scale_logits_by_head_dims
+                assert not (attn_tpl.internal_enable_query_scale and attn_tpl.scale_logits_by_head_dims)
+                assert not attn_tpl.internal_enable_per_dim_scale
+                assert not attn_tpl.scale_query_by_dim_per_head
+                assert not attn_tpl.dconv_qkv
+                assert not attn_tpl.internal_gshard_gaussian_init
+                assert not attn_tpl.use_rotary_position_emb
```

Comment:
Should we remove this assert? Since DotProductAttention can also use RoPE: https://github.com/google/saxml/blob/main/saxml/server/pax/lm/params/lm_cloud.py#L145

Reply:
Offline discussed with Anna and remove the assertion.